http/src/v0/block.rs

```diff
@@ -102,8 +83,41 @@ pub fn put<T: IpfsTypes>(
     path!("block" / "put")
         .and(with_ipfs(ipfs))
         .and(query::<PutQuery>())
-        .and(multipart::form())
-        .and_then(put_query)
+        .and(warp::header::<Mime>("content-type")) // TODO: rejects if missing
+        .and(warp::body::stream())
+        .and_then(inner_put)
+}
+
+async fn inner_put<T: IpfsTypes>(
+    ipfs: Ipfs<T>,
+    opts: PutQuery,
+    mime: Mime,
+    body: impl Stream<Item = Result<impl Buf, warp::Error>> + Unpin,
+) -> Result<impl Reply, Rejection> {
+    let boundary = mime
+        .get_param("boundary")
+        .map(|v| v.to_string())
+        .ok_or_else(|| StringError::from("missing 'boundary' on content-type"))?;
+
+    let buffer = try_only_named_multipart(&["data", "file"], 1024 * 1024, boundary, body).await?;
+
+    // bad thing about Box<[u8]>: converting to it forces an reallocation
```

Comment:
the best we could do here to mitigate this is to delay the boxing to right before the `block` is created

Reply:
This is a good point, thanks. No need to cancel bors, I'll PR this in separatedly.

Reply:
Made #232